### PR TITLE
bpo-46951: Order contents of zipapps

### PR DESCRIPTION
**Pull request by hfinucane**
So that builds are more reproducible[1]. This almost happened for [bpo-30693](https://bugs.python.org/issue30693), but got punted.

[1]: If you want to get all the way there, you need to make sure that file mtimes are clamped, which I've been doing with shell magic. There is a [patch](https://bugs.python.org/issue45344) I haven't tried, that looks promising, floating around on bpo.

https://bugs.python.org/issue46951

### CODE REVIEW

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept this contribution by verifying everyone involved has signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

## Recognized GitHub username

We couldn't find a [bugs.python.org](https://bugs.python.org/) (b.p.o) account corresponding to the following GitHub usernames:

@hfinucane

This might be simply due to a missing "GitHub Name" entry in one's b.p.o account settings. This is necessary for legal reasons before we can look at this contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.




You can [check yourself](https://check-python-cla.herokuapp.com/) to see if the CLA has been received.

Thanks again for the contribution, we look forward to reviewing it!

**Review by serhiy-storchaka — APPROVED** [review]

**Review by raghavthind2005 — APPROVED** [review]

File: Lib/test/test_zipapp.py [hunk 1]
```diff
@@ -54,6 +54,20 @@ def test_create_archive_with_subdirs(self):
...
+        source.mkdir()
+        (source / 'zed.py').touch()
+        (source / 'bin').mkdir()
+        (source / 'bin' / 'qux').touch()
+        (source / 'bin' / 'baz').touch()
+        (source / '__main__.py').touch()
+        target = io.BytesIO()
+        zipapp.create_archive(str(source), target)
+        target.seek(0)
+        self.assertEqual(zipfile.ZipFile(target, 'r').namelist(), sorted(["__main__.py", "bin/", "bin/baz", "bin/qux", "zed.py"]))
```
Comment (serhiy-storchaka) [inline Lib/test/test_zipapp.py]:
```suggestion
        with zipfile.ZipFile(target, 'r') as zf:
            self.assertEqual(zf.namelist(),
                ["__main__.py", "bin/", "bin/baz", "bin/qux", "zed.py"])
```

**Comment by jacksonriley** [comment]
Hi @serhiy-storchaka - it looks like this PR got stalled on CLA Signing, despite it looking like @hfinucane signed the CLA (given that "the-knights-who-say-ni added CLA signed" on 10th March). Do you know if there's a good way to kick the check?

(I was going to open an issue and a PR for this same fix, so was happy to see that there was already one in progress! 😄)

**Review by pfmoore — APPROVED** [review]

**Comment by pfmoore** [comment]
I'm afraid I don't know how to fix the CLA issue here. As it was lack of a bpo username and we've now migrated off bpo, maybe there's no issue any more? I don't know how to confirm this, though (or how to get the CLA check to rerun).

**Comment by pfmoore** [comment]
Seems to have worked, @jacksonriley - thanks!

**Comment by jacksonriley** [comment]
> Seems to have worked, @jacksonriley - thanks!

Nice one :) FWIW I think Serhiy's merge and push would have done the job as well, but all amounts to the same thing!

**Comment by jacksonriley** [comment]
(Looks like just a `make patchcheck` needed to fix the Azure failure, but I don't have permissions)

**Pull request merged**